### PR TITLE
Generalize leader election

### PR DESCRIPTION
**Pull request by C0rWin**
This commit removes assumption of continious relicas ids

Signed-off-by: Artem Barger <bartem@il.ibm.com>

### CODE REVIEW

File: internal/bft/controller.go [hunk 1]
```diff
@@ -108,7 +109,11 @@ func (c *Controller) iAmTheLeader() (bool, uint64) {
 // thread safe
 func (c *Controller) leaderID() uint64 {
 	// TODO use ids order (similar to BFT Smart)
-	return c.getCurrentViewNumber() % c.N
+	nodes := c.Comm.Nodes()
```
Comment (yacovm) [inline internal/bft/controller.go]: why it is important that they are monotonously ascending/descending? I don't think we really use any kind of assumption on their order anywhere...

File: internal/bft/controller.go [hunk 2]
```diff
@@ -108,7 +109,11 @@ func (c *Controller) iAmTheLeader() (bool, uint64) {
...
+	sort.Slice(nodes, func(i, j int) bool {
```
Comment (yacovm) [inline internal/bft/controller.go:113]: why it is important that they are monotonously ascending/descending? I don't think we really use any kind of assumption on their order anywhere...

File: internal/bft/controller.go [hunk 3]
```diff
@@ -108,7 +109,11 @@ func (c *Controller) iAmTheLeader() (bool, uint64) {
 // thread safe
 func (c *Controller) leaderID() uint64 {
 	// TODO use ids order (similar to BFT Smart)
```
Comment (yacovm) [inline internal/bft/controller.go:111]: the TODO can be removed now, no?

File: internal/bft/controller.go (hunk 2 above)
Comment (C0rWin) [reply internal/bft/controller.go:113]: it's important to have it deterministic

File: internal/bft/controller.go (hunk 3 above)
Comment (C0rWin) [reply internal/bft/controller.go:111]: yep

**Comment by HagarMeir** [comment]
from BFT SMaRt:

```
/**
     * Deterministically elects a new leader, based current leader and membership
     * 
     * @return The new leader
     */
    public int getNewLeader() {

        int[] proc = SVController.getCurrentViewProcesses();
        int minProc = proc[0];
        int maxProc = proc[0];
                    
        for (int p : proc) {
            if (p < minProc) minProc = p;
            if (p > maxProc) maxProc = p;
        }
 
        
        do {
            currentLeader++;
            if (currentLeader > maxProc) {

                currentLeader = minProc;    
            }
        } while(!SVController.isCurrentViewMember(currentLeader));
        
        return currentLeader;
    }
```

**Comment by HagarMeir** [comment]
Close this https://github.com/SmartBFT-Go/consensus/issues/152 when finished

**Pull request merged**